node/pkg/watchers/ibc/watcher.go

```diff
@@ -0,0 +1,700 @@
+	}
+	if strings.Contains(string(subResp), "error") {
+		ibcErrors.WithLabelValues("websocket_subscription_error").Inc()
+		return fmt.Errorf("failed to subscribe to events, response: %s", string(subResp))
+	}
+
+	// Start a routine to listen for messages from the contract.
+	common.RunWithScissors(ctx, errC, "ibc_data_pump", func(ctx context.Context) error {
+		return w.handleEvents(ctx, c, errC)
+	})
+
+	// Start a routine to periodically query the wormchain block height.
+	common.RunWithScissors(ctx, errC, "ibc_block_height", func(ctx context.Context) error {
+		return w.handleQueryBlockHeight(ctx, c, errC)
+	})
+
+	// Start a routine for each chain to listen for observation requests.
+	for _, ce := range w.chainMap {
+		common.RunWithScissors(ctx, errC, "ibc_objs_req", func(ctx context.Context) error {
+			return w.handleObservationRequests(ctx, errC, ce)
+		})
+	}
+
+	// Signal to the supervisor that this runnable has finished initialization.
+	supervisor.Signal(ctx, supervisor.SignalHealthy)
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-errC:
+		return err
+	}
+}
+
+// handleEvents reads messages from the IBC receiver contract and processes them.
+func (w *Watcher) handleEvents(ctx context.Context, c *websocket.Conn, errC chan error) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			_, message, err := c.Read(ctx)
+			if err != nil {
+				w.logger.Error("failed to read socket", zap.Error(err))
+				ibcErrors.WithLabelValues("channel_read_error").Inc()
+				return fmt.Errorf("failed to read socket: %w", err)
+			}
+
+			// Received a message from the blockchain.
+			json := string(message)
+
+			txHashRaw := gjson.Get(json, "result.events.tx\\.hash.0")
+			if !txHashRaw.Exists() {
+				w.logger.Warn("message does not have tx hash", zap.String("payload", json))
+				continue
+			}
+			txHash, err := vaa.StringToHash(txHashRaw.String())
+			if err != nil {
+				w.logger.Warn("failed to parse txHash", zap.String("txHash", txHashRaw.String()), zap.Error(err))
+				continue
+			}
+
+			events := gjson.Get(json, "result.data.value.TxResult.result.events")
+			if !events.Exists() {
+				w.logger.Warn("message has no events", zap.String("payload", json))
+				continue
+			}
+
+			for _, event := range events.Array() {
+				if !event.IsObject() {
+					w.logger.Warn("event is invalid", zap.Stringer("tx_hash", txHash), zap.String("event", event.String()))
+					continue
+				}
+				eventType := gjson.Get(event.String(), "type").String()
+				if eventType == "wasm" {
+					evt, err := parseIbcReceivePublishEvent(w.logger, w.contractAddress, event, txHash)
+					if err != nil {
+						w.logger.Error("failed to parse wasm event", zap.Error(err), zap.String("event", event.String()))
+						continue
+					}
+
+					if evt != nil {
+						if err := w.processIbcReceivePublishEvent(txHash, evt, "new"); err != nil {
+							return fmt.Errorf("failed to process new IBC event: %w", err)
```

Comment:
why does this return / restart the routine while the above parsing error just skips?

Reply:
`processIbcReceivePublishEvent` only returns an error if it fails to query the channel id to connection id mapping, which is something that restarting may help?

`parseIbcReceivePublishEvent` can fail based on bad input, and I don't think we want a bad message from an external source to kill the watcher.